repose-aggregator/components/filters/keystone-v2-filter/src/main/scala/org/openrepose/filters/keystonev2/KeystoneV2AuthorizationFilter.scala

```diff
@@ -0,0 +1,110 @@
+/*
+ * _=_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_=
+ * Repose
+ * _-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-
+ * Copyright (C) 2010 - 2015 Rackspace US, Inc.
+ * _-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * =_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_=_
+ */
+package org.openrepose.filters.keystonev2
+
+import javax.inject.{Inject, Named}
+import javax.servlet.ServletRequest
+import javax.servlet.http.HttpServletResponse.SC_INTERNAL_SERVER_ERROR
+
+import org.openrepose.commons.utils.http.OpenStackServiceHeader.TENANT_ID
+import org.openrepose.commons.utils.servlet.http.HttpServletRequestWrapper
+import org.openrepose.core.services.config.ConfigurationService
+import org.openrepose.filters.keystonev2.AbstractKeystoneV2Filter.{KeystoneV2Result, Reject}
+import org.openrepose.filters.keystonev2.KeystoneV2Authorization.{AuthorizationFailed, AuthorizationPassed, doAuthorization}
+import org.openrepose.filters.keystonev2.KeystoneV2Common.{EndpointsData, EndpointsRequestAttributeName, TokenRequestAttributeName, ValidToken}
+import org.openrepose.filters.keystonev2.config.KeystoneV2Config
+
+import scala.util.{Failure, Success, Try}
+
+@Named
+class KeystoneV2AuthorizationFilter @Inject()(configurationService: ConfigurationService)
+  extends AbstractKeystoneV2Filter[KeystoneV2Config](configurationService) {
+
+  import KeystoneV2AuthorizationFilter._
+
+  override val DEFAULT_CONFIG = "keystone-v2-authorization.cfg.xml"
+  override val SCHEMA_LOCATION = "/META-INF/schema/config/keystone-v2-authorization.xsd"
+
+  override val handleFailures: PartialFunction[Try[Unit.type], KeystoneV2Result] = {
+    KeystoneV2Authorization.handleFailures orElse {
+      case Failure(e@(_: MissingTokenException |
+                      _: MissingEndpointsException |
+                      _: InvalidTokenException |
+                      _: InvalidEndpointsException)) =>
+        Reject(SC_INTERNAL_SERVER_ERROR, Some(e.getMessage))
+    }
+  }
+
+  override def doAuth(request: HttpServletRequestWrapper): Try[Unit.type] = {
+    // TODO: Stop using request attributes. Stop using the token. Just use the tenant to roles map, tenant ids, and roles headers.
```

Comment:
This ties back to a previous comment I made, but it's straightforward enough. The one thing I will call out is that, as written, everything is being transported as request attributes right now. I think we all agree that using the headers we are or will be populating for transport is a better idea.

I used request attributes when writing this because we do not yet populate the tenant-to-role map header, so I still needed the token. Since I was already putting the token in a request attribute, I decided to also throw endpoints in a request attribute as kind of a shim. The token attribute cannot be removed until the next story is worked, but the endpoints attribute can be dropped at any time, we will just have to decode and parse the catalog header. As-is, I think we have a MVP, but...

Dropping the endpoints attribute is probably what I'll try to work through today.